File: src/scraper/wikimedia_pageviews.ts
/**
 * Keyless Wikimedia REST Pageviews adapter.
 *
 * Endpoint:
 *   GET /metrics/pageviews/per-article/{project}/{access}/{agent}/{article}/{granularity}/{start}/{end}
 *
 * No API key required. Titles are space→underscore then URI-encoded. Requests
 * use DomainRateLimiter + AbortController timeouts consistent with the scraper.
 */
import { createHash } from "node:crypto";
import { z } from "zod";
import {
  DomainRateLimiter,
  DEFAULT_DOMAIN_LIMITS,
  type DomainLimitConfig,
} from "./rate_limiter";

export const WIKIMEDIA_PAGEVIEWS_BASE =
  "https://wikimedia.org/api/rest_v1/metrics/pageviews/per-article";
export const WIKIMEDIA_PAGEVIEWS_DOMAIN = "wikimedia.org";
export const WIKIMEDIA_PAGEVIEWS_ADAPTER_VERSION = "wikimedia_pageviews_v1";
export const INTELLIGENCE_METRIC_VERSION = "intelligence_metrics_v1";

export const PageviewsAccessSchema = z.enum([
  "all-access",
  "desktop",
  "mobile-app",
  "mobile-web",
]);
export const PageviewsAgentSchema = z.enum([
  "all-agents",
  "user",
  "spider",
  "automated",
]);
export const PageviewsGranularitySchema = z.enum(["daily", "monthly"]);

export type PageviewsAccess = z.infer<typeof PageviewsAccessSchema>;
export type PageviewsAgent = z.infer<typeof PageviewsAgentSchema>;
export type PageviewsGranularity = z.infer<typeof PageviewsGranularitySchema>;

const PageviewItemSchema = z.object({
  project: z.string(),
  article: z.string(),
  granularity: z.string(),
  timestamp: z.string(),
  access: z.string(),
  agent: z.string(),
  views: z.number(),
});

export const PageviewsResponseSchema = z.object({
  items: z.array(PageviewItemSchema).min(1),
});

export type PageviewsResponse = z.infer<typeof PageviewsResponseSchema>;
export type PageviewItem = z.infer<typeof PageviewItemSchema>;

export type WikimediaPageviewsRequest = {
  /** Article title as humans write it (spaces allowed). */
  articleTitle: string;
  project?: string;
  access?: PageviewsAccess;
  agent?: PageviewsAgent;
  granularity?: PageviewsGranularity;
  /** YYYYMMDD or YYYYMMDDHH */
  start: string;
  /** YYYYMMDD or YYYYMMDDHH */
  end: string;
  artistKey?: string;
  festivalKey?: string;
  editionKey?: string;
  editionYear?: number;
};

export type WikimediaPageviewsResult = {
  ok: boolean;
  status: "ok" | "error" | "missing";
  httpStatus: number;
  sourceUrl: string;
  retrievedAt: string;
  project: string;
  access: PageviewsAccess;
  agent: PageviewsAgent;
  articleTitle: string;
  encodedArticle: string;
  granularity: PageviewsGranularity;
  start: string;
  end: string;
  items: PageviewItem[];
  valueSum: number | null;
  rawResponse: unknown;
  errorCode?: string;
  errorMessage?: string;
  adapterVersion: string;
  metricVersion: string;
  provenance: {
    sourceSystem: "wikimedia";
    endpoint: "per-article";
    userAgent: string;
    request: Record<string, string | number | undefined>;
  };
};

export type WikimediaPageviewsClientOptions = {
  fetchImpl?: typeof fetch;
  rateLimiter?: DomainRateLimiter;
  timeoutMs?: number;
  userAgent?: string;
  baseUrl?: string;
  now?: () => Date;
  domainLimits?: Partial<DomainLimitConfig>;
};

const DEFAULT_USER_AGENT =
  "FestivalBloomberg/0.1 (intelligence-metrics; keyless-pageviews; contact=festival-bloomberg)";

/**
 * Encode a Wikipedia article title for the per-article path segment.
 * Spaces become underscores, then URI-encoding is applied once.
 */
export function encodePageviewsArticleTitle(title: string): string {
  const normalized = title.trim().replace(/\s+/g, "_");
  if (!normalized) {
    throw new Error("article_title_empty");
  }
  return encodeURIComponent(normalized);
}

export function buildPageviewsUrl(input: {
  project: string;
  access: PageviewsAccess;
  agent: PageviewsAgent;
  articleTitle: string;
  granularity: PageviewsGranularity;
  start: string;
  end: string;
  baseUrl?: string;
}): { url: string; encodedArticle: string } {
  const encodedArticle = encodePageviewsArticleTitle(input.articleTitle);
  const project = input.project.replace(/\/+$/, "");
  const base = (input.baseUrl ?? WIKIMEDIA_PAGEVIEWS_BASE).replace(/\/+$/, "");
  const url = [
    base,
    encodeURIComponent(project),
    encodeURIComponent(input.access),
    encodeURIComponent(input.agent),
    encodedArticle,
    encodeURIComponent(input.granularity),
    encodeURIComponent(input.start),
    encodeURIComponent(input.end),
  ].join("/");
  return { url, encodedArticle };
}

export function parsePageviewsResponse(payload: unknown): {
  items: PageviewItem[];
  valueSum: number;
} {
  const parsed = PageviewsResponseSchema.parse(payload);
  const valueSum = parsed.items.reduce((sum, item) => sum + item.views, 0);
  return { items: parsed.items, valueSum };
}

export function attentionObservationKey(parts: {
  artistKey: string;
  sourceSystem: string;
  metricKind: string;
  project?: string | null;
  periodStart?: string | null;
  periodEnd?: string | null;
  metricVersion: string;
}): string {
  const material = [
    parts.artistKey,
    parts.sourceSystem,
    parts.metricKind,
    parts.project ?? "",
    parts.periodStart ?? "",
    parts.periodEnd ?? "",
    parts.metricVersion,
  ].join("|");
  return createHash("sha256").update(material).digest("hex").slice(0, 32);
}

function yyyymmddToIsoDate(raw: string): string {
  const digits = raw.replace(/\D/g, "").slice(0, 8);
  if (digits.length !== 8) {
    throw new Error(`invalid_pageviews_date:${raw}`);
  }
  return `${digits.slice(0, 4)}-${digits.slice(4, 6)}-${digits.slice(6, 8)}`;
}

export class WikimediaPageviewsClient {
  private readonly fetchImpl: typeof fetch;
  private readonly rateLimiter: DomainRateLimiter;
  private readonly timeoutMs: number;
  private readonly userAgent: string;
  private readonly baseUrl: string;
  private readonly now: () => Date;

  constructor(opts: WikimediaPageviewsClientOptions = {}) {
    this.fetchImpl = opts.fetchImpl ?? fetch;
    this.timeoutMs = opts.timeoutMs ?? 15_000;
    this.userAgent = opts.userAgent ?? DEFAULT_USER_AGENT;
    this.baseUrl = opts.baseUrl ?? WIKIMEDIA_PAGEVIEWS_BASE;
    this.now = opts.now ?? (() => new Date());
    this.rateLimiter =
      opts.rateLimiter ??
      new DomainRateLimiter({
        defaults: {
          ...DEFAULT_DOMAIN_LIMITS,
          tokensPerSecond: 2,
          bucketSize: 4,
          minSpacingMs: 200,
          ...opts.domainLimits,
        },
      });
  }

  async fetchPerArticle(
    request: WikimediaPageviewsRequest,
  ): Promise<WikimediaPageviewsResult> {
    const project = request.project ?? "en.wikipedia";
    const access = request.access ?? "all-access";
    const agent = request.agent ?? "user";
    const granularity = request.granularity ?? "daily";
    const retrievedAt = this.now().toISOString();
    const { url, encodedArticle } = buildPageviewsUrl({
      project,
      access,
      agent,
      articleTitle: request.articleTitle,
      granularity,
      start: request.start,
      end: request.end,
      baseUrl: this.baseUrl,
    });

    const provenance: WikimediaPageviewsResult["provenance"] = {
      sourceSystem: "wikimedia",
      endpoint: "per-article",
      userAgent: this.userAgent,
      request: {
        project,
        access,
        agent,
        articleTitle: request.articleTitle,
        encodedArticle,
        granularity,
        start: request.start,
        end: request.end,
        artistKey: request.artistKey,
        festivalKey: request.festivalKey,
        editionKey: request.editionKey,
        editionYear: request.editionYear,
      },
    };

    const baseResult: Omit<
      WikimediaPageviewsResult,
      "ok" | "status" | "httpStatus" | "items" | "valueSum" | "rawResponse"
    > & {
      items: PageviewItem[];
      valueSum: number | null;
      rawResponse: unknown;
    } = {
      sourceUrl: url,
      retrievedAt,
      project,
      access,
      agent,
      articleTitle: request.articleTitle.trim(),
      encodedArticle,
      granularity,
      start: request.start,
      end: request.end,
      items: [],
      valueSum: null,
      rawResponse: null,
      adapterVersion: WIKIMEDIA_PAGEVIEWS_ADAPTER_VERSION,
      metricVersion: INTELLIGENCE_METRIC_VERSION,
      provenance,
    };

    try {
      const response = await this.rateLimiter.executeWithRetry(
        WIKIMEDIA_PAGEVIEWS_DOMAIN,
        async () => this.requestOnce(url),
      );

      if (response.httpStatus === 404) {
        return {
          ...baseResult,
          ok: false,
          status: "missing",
          httpStatus: 404,
          rawResponse: response.body,
          errorCode: "pageviews_not_found",
          errorMessage: "Article or date range not found in Pageviews API",
        };
      }

      if (!response.ok) {
        return {
          ...baseResult,
          ok: false,
          status: "error",
          httpStatus: response.httpStatus,
          rawResponse: response.body,
          errorCode: `http_${response.httpStatus}`,
          errorMessage: response.errorMessage ?? `HTTP ${response.httpStatus}`,
        };
      }

      try {
        const { items, valueSum } = parsePageviewsResponse(response.body);
        return {
          ...baseResult,
          ok: true,
          status: "ok",
          httpStatus: response.httpStatus,
          items,
          valueSum,
          rawResponse: response.body,
        };
      } catch (error) {
        return {
          ...baseResult,
          ok: false,
          status: "error",
          httpStatus: response.httpStatus,
          rawResponse: response.body,
          errorCode: "response_parse_error",
          errorMessage: error instanceof Error ? error.message : String(error),
        };
      }
    } catch (error) {
      const status =
        typeof (error as { status?: number })?.status === "number"
          ? (error as { status: number }).status
          : 0;
      return {
        ...baseResult,
        ok: false,
        status: "error",
        httpStatus: status,
        rawResponse: null,
        errorCode:
          (error as { code?: string })?.code ??
          (status ? `http_${status}` : "request_failed"),
        errorMessage: error instanceof Error ? error.message : String(error),
      };
    }
  }

  /**
   * Map a successful/failed fetch into a normalized attention observation row.
   */
  toAttentionObservation(
    result: WikimediaPageviewsResult,
    identity: {
      artistKey: string;
      festivalKey?: string;
      editionKey?: string;
      editionYear?: number;
    },
  ) {
    let periodStart: string | null = null;
    let periodEnd: string | null = null;
    try {
      periodStart = yyyymmddToIsoDate(result.start);
      periodEnd = yyyymmddToIsoDate(result.end);
    } catch {
      periodStart = null;
      periodEnd = null;
    }

    const observationKey = attentionObservationKey({
      artistKey: identity.artistKey,
      sourceSystem: "wikimedia",
      metricKind: "pageviews",
      project: result.project,
      periodStart,
      periodEnd,
      metricVersion: result.metricVersion,
    });

    return {
      observation_key: observationKey,
      artist_key: identity.artistKey,
      festival_key: identity.festivalKey ?? null,
      edition_key: identity.editionKey ?? null,
      edition_year: identity.editionYear ?? null,
      source_system: "wikimedia" as const,
      metric_kind: "pageviews" as const,
      project: result.project,
      access_method: result.access,
      agent: result.agent,
      article_title: result.articleTitle,
      granularity: result.granularity,
      period_start: periodStart,
      period_end: periodEnd,
      value: result.valueSum,
      value_sum: result.valueSum,
      value_unit: "pageviews",
      status: result.status,
      error_code: result.errorCode ?? null,
      error_message: result.errorMessage ?? null,
      source_url: result.sourceUrl,
      retrieved_at: result.retrievedAt,
      raw_response_json: result.rawResponse,
      provenance_json: result.provenance,
      metric_version: result.metricVersion,
    };
  }

  private async requestOnce(url: string): Promise<{
    ok: boolean;
    httpStatus: number;
    body: unknown;
    errorMessage?: string;
  }> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await this.fetchImpl(url, {
        method: "GET",
        headers: {
          Accept: "application/json",
          "User-Agent": this.userAgent,
        },
        signal: controller.signal,
      });

      const text = await response.text();
      let body: unknown = text;
      if (text) {
        try {
          body = JSON.parse(text);
        } catch {
          body = text;
        }
      } else {
        body = null;
      }

      if (!response.ok) {
        const retryable =
          response.status === 429 || response.status >= 500;
        if (retryable) {
          const err = Object.assign(
            new Error(`Wikimedia Pageviews HTTP ${response.status}`),
            {
              status: response.status,
              retryAfter: response.headers.get("retry-after"),
              code: `http_${response.status}`,
            },
          );
          throw err;
        }
      }

      return {
        ok: response.ok,
        httpStatus: response.status,
        body,
        errorMessage: response.ok
          ? undefined
          : `Wikimedia Pageviews HTTP ${response.status}`,
      };
    } catch (error) {
      if ((error as { name?: string })?.name === "AbortError") {
        const err = Object.assign(new Error("Wikimedia Pageviews timeout"), {
          status: 0,
          code: "timeout",
        });
        throw err;
      }
      throw error;
    } finally {
      clearTimeout(timer);
    }
  }
}

File: python/festival_bloomberg/attention/wikimedia_pageviews.py
"""Key-free Wikimedia REST Pageviews collector (attention time series).

Endpoint:
    GET /metrics/pageviews/per-article/{project}/{access}/{agent}/{article}/{granularity}/{start}/{end}

No API key is required. Titles are space -> underscore then URI-encoded. Each
per-article response is persisted as one ``metrics.artist_attention_observations``
row keyed by a stable hash, so re-running a window is idempotent and never
rewrites history.

Semantics preserved:

- Pageviews are an ATTENTION channel (``ATTENTION_CONSUMPTION_SAMPLE``), never
  LOCAL_DEMAND and never a ticket-demand proxy.
- The article title is stored verbatim (``article_title``); a 404 title is
  persisted with ``status = 'missing'`` — UNKNOWN is never encoded as a zero
  value.
- ``period_start`` / ``period_end`` are the *observation window*; ``retrieved_at``
  is when we fetched it. They are never collapsed.
"""

from __future__ import annotations

import hashlib
import json
import time
import urllib.parse
from datetime import date, datetime, timedelta, timezone
from typing import Any

from ..identity.spotify import normalize_name

WIKIMEDIA_PAGEVIEWS_BASE = "https://wikimedia.org/api/rest_v1/metrics/pageviews/per-article"
METRIC_VERSION = "wikimedia_pageviews_v1"
SOURCE_SYSTEM = "wikimedia"

#: The Analytics API serves pageview data starting on this day. Days before
#: this are OUTSIDE the source's existence and are therefore UNAVAILABLE —
#: never MISSING, never ZERO. See the Wikimedia Analytics API reference:
#: "These endpoints serve data starting on July 1, 2015."
WIKIMEDIA_SERIES_START = date(2015, 7, 1)

#: Availability policy version for the daily pageview aggregate. Bump when the
#: availability semantics below change.
WIKIMEDIA_AVAILABILITY_POLICY_VERSION = "wikimedia_pageviews_availability_v1"


def wikimedia_available_at(observation_day: date) -> date:
    """Day-level publication bound for a Wikimedia daily pageview aggregate.

    Wikimedia loads a day's pageview data at the end of the relevant period:
    the aggregate for ``observation_day`` D becomes knowable from the source at
    D+1 (00:00 UTC). This is a conservative availability bound derived from the
    source's documented load semantics — it is NOT the retrieval time, and it
    is NOT the observation day itself.

    PIT admissibility for a window ending at cutoff therefore requires
    ``observation_day < cutoff AND wikimedia_available_at(observation_day) <
    cutoff``. ``retrieved_at`` (when Festival Bloomberg happened to download
    the value) is provenance and is NEVER an admissibility gate.
    """
    return observation_day + timedelta(days=1)


def artist_key_for(name: str) -> str:
    """Documented fallback artist_key: ``name::<normalized_name>``."""
    return f"name::{normalize_name(name)}"


def encode_title(title: str) -> str:
    normalized = (title or "").strip().replace(" ", "_")
    if not normalized:
        raise ValueError("article_title_empty")
    return urllib.parse.quote(normalized)


def build_pageviews_url(
    title: str,
    *,
    project: str = "en.wikipedia",
    access: str = "all-access",
    agent: str = "user",
    granularity: str = "daily",
    start: str,
    end: str,
    base_url: str = WIKIMEDIA_PAGEVIEWS_BASE,
) -> str:
    encoded = encode_title(title)
    parts = [base_url.rstrip("/"), project, access, agent, encoded, granularity, start, end]
    return "/".join(parts)


def pageviews_observation_key(
    *,
    artist_key: str,
    project: str,
    period_start: str | None,
    period_end: str | None,
) -> str:
    material = "|".join(
        [artist_key, SOURCE_SYSTEM, "pageviews", project, period_start or "", period_end or "", METRIC_VERSION]
    )
    return hashlib.sha256(material.encode("utf-8")).hexdigest()[:32]


def _yyyymmdd_to_iso(raw: str) -> str | None:
    digits = "".join(ch for ch in (raw or "") if ch.isdigit())[:8]
    if len(digits) != 8:
        return None
    return f"{digits[:4]}-{digits[4:6]}-{digits[6:8]}"


def build_pageviews_observation(
    *,
    artist_name: str,
    title: str,
    project: str,
    access: str,
    agent: str,
    granularity: str,
    start: str,
    end: str,
    items: list[dict[str, Any]],
    status: str,
    error_code: str | None,
    error_message: str | None,
    source_url: str,
    retrieved_at: str,
    raw_response: Any = None,
) -> dict[str, Any]:
    artist_key = artist_key_for(artist_name)
    period_start = _yyyymmdd_to_iso(start)
    period_end = _yyyymmdd_to_iso(end)
    value_sum = None
    if status == "ok":
        value_sum = sum(int(i.get("views") or 0) for i in items)
    return {
        "observation_key": pageviews_observation_key(
            artist_key=artist_key, project=project,
            period_start=period_start, period_end=period_end,
        ),
        "artist_key": artist_key,
        "festival_key": None,
        "edition_key": None,
        "edition_year": None,
        "source_system": SOURCE_SYSTEM,
        "metric_kind": "pageviews",
        "project": project,
        "access_method": access,
        "agent": agent,
        "article_title": title.strip(),
        "granularity": granularity,
        "period_start": period_start,
        "period_end": period_end,
        "value": value_sum,
        "value_sum": value_sum,
        "value_unit": "pageviews",
        "status": status,
        "error_code": error_code,
        "error_message": error_message,
        "source_url": source_url,
        "retrieved_at": retrieved_at,
        "raw_response_json": json.dumps(raw_response, default=str) if raw_response is not None else None,
        "provenance_json": json.dumps(
            {
                "source_system": SOURCE_SYSTEM,
                "endpoint": "per-article",
                "project": project,
                "access": access,
                "agent": agent,
                "granularity": granularity,
                "start": start,
                "end": end,
                "artist_name": artist_name,
                "article_title": title.strip(),
            },
            default=str,
        ),
        "metric_version": METRIC_VERSION,
    }


def persist_pageviews(conn, row: dict[str, Any]) -> int:
    """Insert one pageviews observation (idempotent by observation_key)."""
    exists = conn.execute(
        "SELECT 1 FROM metrics.artist_attention_observations WHERE observation_key = ?",
        [row["observation_key"]],
    ).fetchone()
    if exists:
        return 0
    conn.execute(
        """
        INSERT INTO metrics.artist_attention_observations
            (observation_key, artist_key, festival_key, edition_key, edition_year,
             source_system, metric_kind, project, access_method, agent, article_title,
             granularity, period_start, period_end, value, value_sum, value_unit,
             status, error_code, error_message, source_url, retrieved_at,
             raw_response_json, provenance_json, metric_version, ingested_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP)
        """,
        [
            row["observation_key"], row["artist_key"], row["festival_key"], row["edition_key"],
            row["edition_year"], row["source_system"], row["metric_kind"], row["project"],
            row["access_method"], row["agent"], row["article_title"], row["granularity"],
            row["period_start"], row["period_end"], row["value"], row["value_sum"],
            row["value_unit"], row["status"], row["error_code"], row["error_message"],
            row["source_url"], row["retrieved_at"], row["raw_response_json"],
            row["provenance_json"], row["metric_version"],
        ],
    )
    return 1


def fetch_pageviews(
    transport,
    *,
    title: str,
    start: str,
    end: str,
    project: str = "en.wikipedia",
    access: str = "all-access",
    agent: str = "user",
    granularity: str = "daily",
    timeout_seconds: float = 30.0,
) -> dict[str, Any]:
    """Fetch a per-article pageviews window and normalize it (no persistence)."""
    from ..acquisition.transport import TransportError

    url = build_pageviews_url(
        title, project=project, access=access, agent=agent,
        granularity=granularity, start=start, end=end,
    )
    retrieved_at = datetime.now(timezone.utc).isoformat()
    base = dict(
        project=project, access=access, agent=agent, granularity=granularity,
        start=start, end=end, source_url=url, retrieved_at=retrieved_at,
        items=[], raw_response=None,
    )
    try:
        response = transport.request("GET", url, headers={"Accept": "application/json"}, timeout_seconds=timeout_seconds)
    except TransportError as exc:
        return {**base, "status": "error", "error_code": "network", "error_message": str(exc)}

    if response.status == 404:
        return {**base, "status": "missing", "error_code": "pageviews_not_found",
                "error_message": "Article or date range not found", "raw_response": response.body.decode("utf-8", "replace")}
    if response.status != 200:
        return {**base, "status": "error", "error_code": f"http_{response.status}",
                "error_message": f"HTTP {response.status}", "raw_response": response.body.decode("utf-8", "replace")}
    try:
        payload = json.loads(response.body.decode("utf-8"))
    except (ValueError, UnicodeDecodeError):
        return {**base, "status": "error", "error_code": "response_parse_error",
                "error_message": "non-JSON response", "raw_response": response.body.decode("utf-8", "replace")}
    items = (payload or {}).get("items") or []
    if not items:
        return {**base, "status": "missing", "error_code": "pageviews_empty",
                "error_message": "No pageview items returned", "raw_response": payload}
    return {**base, "status": "ok", "items": items, "raw_response": payload}


def collect_artist_pageviews(
    conn,
    transport,
    *,
    names: list[str],
    days: int = 30,
    project: str = "en.wikipedia",
    access: str = "all-access",
    agent: str = "user",
    min_interval_seconds: float = 0.5,
) -> dict[str, Any]:
    """Bounded pageviews collection over a list of artist names.

    ``names`` are raw display names; the article title defaults to the name
    (the same conservative fallback as the TypeScript scraper). A 404 title
    is persisted as ``missing``, never as a fabricated zero.
    """
    end_dt = datetime.now(timezone.utc).date()
    start_dt = end_dt - timedelta(days=max(1, days))
    start = start_dt.strftime("%Y%m%d")
    end = end_dt.strftime("%Y%m%d")
    summary = {
        "status": "RUNNING",
        "names_attempted": 0,
        "ok": 0,
        "missing": 0,
        "error": 0,
        "rows_persisted": 0,
        "window_start": start,
        "window_end": end,
    }
    for index, name in enumerate(names):
        if not name or not name.strip():
            continue
        if index and min_interval_seconds > 0:
            time.sleep(min_interval_seconds)
        summary["names_attempted"] += 1
        result = fetch_pageviews(
            transport, title=name, start=start, end=end,
            project=project, access=access, agent=agent,
        )
        row = build_pageviews_observation(
            artist_name=name,
            title=name,
            project=result["project"],
            access=result["access"],
            agent=result["agent"],
            granularity=result["granularity"],
            start=result["start"],
            end=result["end"],
            items=result["items"],
            status=result["status"],
            error_code=result.get("error_code"),
            error_message=result.get("error_message"),
            source_url=result["source_url"],
            retrieved_at=result["retrieved_at"],
            raw_response=result.get("raw_response"),
        )
        summary[result["status"]] += 1
        summary["rows_persisted"] += persist_pageviews(conn, row)
    summary["status"] = "COMPLETE"
    return summary
